server/src/streaming/deduplication/message_deduplicator.rs

```diff
@@ -0,0 +1,83 @@
+use moka::future::Cache;
+use std::time::Duration;
+
+#[derive(Debug)]
+pub struct MessageDeduplicator {
+    cache: Cache<u128, bool>,
+}
+
+impl MessageDeduplicator {
+    /// Creates a new message deduplicator with the given max entries and time to live for each ID.
+    pub fn new(max_entries: Option<u64>, ttl: Option<Duration>) -> Self {
+        let mut cache = Cache::builder();
+        if let Some(max_entries) = max_entries {
+            cache = cache.max_capacity(max_entries);
+        }
+        if let Some(ttl) = ttl {
+            cache = cache.time_to_live(ttl);
+        }
+
+        Self {
+            cache: cache.build(),
+        }
+    }
+
+    /// Checks if the given ID exists.
+    pub fn exists(&self, id: u128) -> bool {
+        self.cache.contains_key(&id)
+    }
+
+    /// Inserts the given ID.
+    pub async fn insert(&self, id: u128) {
+        self.cache.insert(id, true).await;
+    }
+
+    /// Tries to insert the given ID, returns false if it already exists.
+    pub async fn try_insert(&self, id: u128) -> bool {
```

Comment:
perhaps you could try to implement this with`try_get_with_by_ref`?
https://docs.rs/moka/latest/moka/future/struct.Cache.html#method.try_get_with_by_ref

Reply:
Done